[no-redeclare] ignoreDeclarationMerge when using enum+namespace

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```ts
enum E {
  A = "a",
  B = "b"
}

namespace E {
  export function values(): string[] {
    return [E.A, E.B];
  }
}
```

```JSON
{
  "rules": {
     "@typescript-eslint/no-redeclare": ["error", { "ignoreDeclarationMerge": true }],
  }
}
```

**Expected Result**

I'd expect the rule not to emit an error in this case.

**Actual Result**

```
 6:11  error  'E' is already defined                                                                @typescript-eslint/no-redeclare
```

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.27.0` |
| `@typescript-eslint/parser`        | `4.27.0` |
| `TypeScript`                       | `4.3.0` |
| `ESLint`                           | `7.29.0` |
| `node`                             | `14.16.1` |

--- a/packages/eslint-plugin/docs/rules/no-redeclare.md
+++ b/packages/eslint-plugin/docs/rules/no-redeclare.md
@@ -41,6 +41,7 @@ When set to `true`, the rule will ignore declaration merges between the followin
 - class + namespace
 - class + interface + namespace
 - function + namespace
+- enum + namespace
 
 Examples of **correct** code with `{ ignoreDeclarationMerge: true }`:
 

--- a/packages/eslint-plugin/src/rules/no-redeclare.ts
+++ b/packages/eslint-plugin/src/rules/no-redeclare.ts
@@ -63,6 +63,10 @@ export default util.createRule<Options, MessageIds>({
       AST_NODE_TYPES.TSModuleDeclaration,
       AST_NODE_TYPES.FunctionDeclaration,
     ]);
+    const ENUM_DECLARATION_MERGE_NODES = new Set<AST_NODE_TYPES>([
+      AST_NODE_TYPES.TSEnumDeclaration,
+      AST_NODE_TYPES.TSModuleDeclaration,
+    ]);
 
     function* iterateDeclarations(variable: TSESLint.Scope.Variable): Generator<
       {
@@ -164,8 +168,29 @@ export default util.createRule<Options, MessageIds>({
             return;
           }
 
-          // there's more than one class declaration, which needs to be reported
+          // there's more than one function declaration, which needs to be reported
           for (const { identifier } of functionDecls) {
+            yield { type: 'syntax', node: identifier, loc: identifier.loc };
+          }
+          return;
+        }
+
+        if (
+          // enum + namespace merging
+          identifiers.every(({ parent }) =>
+            ENUM_DECLARATION_MERGE_NODES.has(parent.type),
+          )
+        ) {
+          const enumDecls = identifiers.filter(
+            ({ parent }) => parent.type === AST_NODE_TYPES.TSEnumDeclaration,
+          );
+          if (enumDecls.length === 1) {
+            // safe declaration merging
+            return;
+          }
+
+          // there's more than one enum declaration, which needs to be reported
+          for (const { identifier } of enumDecls) {
             yield { type: 'syntax', node: identifier, loc: identifier.loc };
           }
           return;

--- a/packages/eslint-plugin/tests/rules/no-redeclare.test.ts
+++ b/packages/eslint-plugin/tests/rules/no-redeclare.test.ts
@@ -128,6 +128,13 @@ namespace A {}
     },
     {
       code: `
+enum A {}
+namespace A {}
+      `,
+      options: [{ ignoreDeclarationMerge: true }],
+    },
+    {
+      code: `
 function A() {}
 namespace A {}
       `,
@@ -600,6 +607,23 @@ class A {}
             id: 'A',
           },
           line: 3,
+        },
+      ],
+    },
+    {
+      code: `
+enum A {}
+namespace A {}
+enum A {}
+      `,
+      options: [{ ignoreDeclarationMerge: true }],
+      errors: [
+        {
+          messageId: 'redeclared',
+          data: {
+            id: 'A',
+          },
+          line: 4,
         },
       ],
     },

